### PR TITLE
Replace checkNotTimeout with raceTimer helper

### PR DESCRIPTION
**Pull request by arinwt**
`checkNotTimeout` was a confusing helper function to use in conjunction with timer results. This PR replaces it with an overall better solution: `raceTimer` is a function that lets you pass the two promises in and return when either of them complete.

### CODE REVIEW

File: packages/runtime/container-runtime/src/summaryGenerator.ts [hunk 1]
```diff
@@ -14,12 +14,28 @@ import {
 } from "./summarizerTypes";
 import { IClientSummaryWatcher } from "./summaryCollection";
 
-export const checkNotTimeout = <T>(something: T | IPromiseTimerResult | undefined): something is T => {
-    if (something === undefined) {
-        return false;
-    }
-    return (something as IPromiseTimerResult).timerResult === undefined;
+export type RaceTimerResult<T> = {
+    readonly result: IPromiseTimerResult["timerResult"];
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/summaryGenerator.ts:18]: Would it be a bit more easier to follow the code if we delete IPromiseTimerResult and instead use something like
type TimerResult = "timeout" | "cancel"?
Also any chance you can add comments above IPromiseTimerResult  describing how everybody should think about what timeout and cancel mean?

File: packages/runtime/container-runtime/src/summaryGenerator.ts [hunk 2]
```diff
@@ -14,12 +14,28 @@ import {
...
+    readonly result: "done";
+    readonly value: T;
 };
+export const raceTimer = async <T>(
+    p: Promise<T>,
+    timer: Promise<IPromiseTimerResult>,
+): Promise<RaceTimerResult<T>> => Promise.race([
+    p.then(
+        (value) => ({ result: "done", value } as const),
+        (error) => ({ result: "timeoutError", error } as const),
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/summaryGenerator.ts:32]: I'm a bit confused about "timeoutError". It's not timeout error, it's just an error, no?
And the error case in timer is timeoutError

File: packages/runtime/container-runtime/src/summaryGenerator.ts [hunk 3]
```diff
@@ -14,12 +14,28 @@ import {
...
+    ),
+    timer.then(
+        (result) => ({ result: result.timerResult } as const),
+        (error) => ({ result: "error", error } as const),
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/summaryGenerator.ts:36]: What are the cases when this error can happen? If this is not expected result (i.e. other than logic error in small amount of code, we do not expect to hit it), I'd rather use fail-fast approach and fail in some way that does not expose this kind of path to consumers. I.e. either log and ignore, or roll it into error path for p, or make it unhandled promise - something that the caller does not need to think about.

Comment (vladsud) [reply packages/runtime/container-runtime/src/summaryGenerator.ts:36]: You can actually see it from usage - there is only checking for "done" state, the rest of the code assumes that not done means timeout. Which is not entirely correct, as likely we want to differentiate error (not timer error) from timeout

Comment (arinwt) [reply packages/runtime/container-runtime/src/summaryGenerator.ts:36]: Right, I was just trying to propagate max info out of this function and let caller decide what to do- fail fast. I'll rethink this.

File: packages/runtime/container-runtime/src/summaryGenerator.ts (hunk 2 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/summaryGenerator.ts:32]: Oh I had these reversed, nice catch!

File: packages/runtime/container-runtime/src/summaryGenerator.ts (hunk 1 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/summaryGenerator.ts:18]: Yes, but those are common-utils changes, which require compatibility iterations. I think there are other changes/improvements to bundle within timer.ts.

File: packages/runtime/container-runtime/src/summaryGenerator.ts (hunk 3 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/summaryGenerator.ts:36]: I added different errors.. let me know if you think they are any better.

Comment (arinwt) [reply packages/runtime/container-runtime/src/summaryGenerator.ts]: Okay I have left them as promise rejections now. This way they will be sweeped up by parent catches and logged as "unexpected" errors in the corresponding places: SummaryManager and RunningSummarizer.

Comment (arinwt) [reply packages/runtime/container-runtime/src/summaryGenerator.ts]: @vladsud - bump!

**Review by vladsud — APPROVED** [review]

**Pull request merged**